monitor for creation of strings with distinctive memory corruption pattern

josharian opened the pull request:
Not meant to be merged.

This the first, easiest, least invasive check. The goal here is to discover corruption earlier, to aid in reproducibility; it is unlikely to provide a useful diagnosis of the corruption itself.

For tailscale/tailscale#921

josharian committed:
runtime: monitor for creation of strings with distinctive memory corruption pattern

Change-Id: I7270f8ec5198454e91200e299f1f75ff99689feb

bradfitz commented:
You meant `For tailscale/tailscale#921`

bradfitz commented:
Instead of a throw, can we make it print a stack trace (including the address of the &byteslice[0]) but keep on running without crashing?

josharian committed:
cmd/compile: when envvar TROUBLE=1 is set, print possible trouble types

Be sure to compile with -a to bypass the cmd/go cache, because envvar.

Lots of duplicate lines in the output. Use with uniq or pct.

Change-Id: I5a3db2e9951c8f132bb970320db6f07069cae6f2

josharian commented:
> Instead of a throw, can we make it print a stack trace (including the address of the &byteslice[0]) but keep on running without crashing?

Done.

Also, if you do `TROUBLE=1 go build -a pkg`, it'll log all 16 byte types with padding/blank field at bytes [4,7).

bradfitz commented on src/cmd/compile/internal/gc/ssa.go:
```diff
@@ -1010,17 +1010,52 @@ func (s *state) rawLoad(t *types.Type, src *ssa.Value) *ssa.Value {
 
 func (s *state) store(t *types.Type, dst, val *ssa.Value) {
+	if isTroubleType(t) {
+		fmt.Printf("storing padded 4-4-8 type: %v\n", t)
```
this should be stderr (and below) so redo doesn't consider it part of the file output probably

bradfitz commented:
Fancy!

bradfitz commented on src/runtime/string.go:
```diff
@@ -102,6 +102,16 @@ func slicebytetostring(buf *tmpBuf, ptr *byte, n int) (str string) {
 		stringStructOf(&str).len = 1
 		return
 	}
+	if n >= 16 {
```
Aside: I guess this doesn't catch strings.Builder conversions?
